egg-herbie/main.rkt

```diff
@@ -60,6 +60,92 @@
 ; Checks for a condition on MacOS if x86 Racket is being used on an ARM mac.
 (define-ffi-definer define-eggmath (ffi-lib libeggmath-path #:fail handle-eggmath-import-failure))
 
+;; Frees a Rust-allocated C-string
+(define-eggmath destroy_string (_fun _pointer -> _void))
+
+;; Gets the length of a Rust-allocated C-string in bytes,
+;; excluding the nul terminator.
+(define-eggmath string_length (_fun _pointer -> _uint32))
+
+;; Converts a Racket string to a C-style string.
+(define (string->_rust/string s #:raw? [raw? #f])
+  (define bstr (string->bytes/utf-8 s))
+  (define n (bytes-length bstr))
+  (define p (malloc (if raw? 'raw 'atomic) (add1 n)))
+  (memcpy p bstr n)
+  (ptr-set! p _byte n 0)
+  p)
```

Comment:
I do think that Rust's native string type is length+data pointer, not a C-string with a null terminator, so it's possible we could avoid the copy if we were clever.

Reply:
```
#[derive(PartialEq, PartialOrd, Eq, Ord)]
#[stable(feature = "rust1", since = "1.0.0")]
#[cfg_attr(not(test), lang = "String")]
pub struct String {
    vec: Vec<u8>,
}
```

Reply:
This gives me some ideas. Mainly, `u8vector` from `ffi/vector`.

Reply:
Ok I'll save that for a follow-up PR. I briefly skimmed the docs, essentially `String` has `as_ptr()` which returns a `u8` pointer which makes the conversion to a byte string straightforward. The only hitch is we compute the length of the string on the Rust side by sending the pointer back over (and I don't see a method that goes from `*const u8` to `String`).

Reply:
Another PR sounds right.